File: app/src/main/java/com/noxvision/app/ui/hunting/WeatherScreen.kt
package com.noxvision.app.ui.hunting

import android.Manifest
import android.content.pm.PackageManager
import androidx.activity.compose.rememberLauncherForActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.content.ContextCompat
import com.noxvision.app.hunting.database.entities.CachedWeather
import com.noxvision.app.hunting.location.HuntingLocationManager
import com.noxvision.app.hunting.moon.MoonPhaseCalculator
import com.noxvision.app.hunting.weather.WeatherCache
import com.noxvision.app.hunting.weather.WeatherIconHelper
import com.noxvision.app.ui.NightColors
import com.noxvision.app.ui.SettingsSectionHeader
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.*

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun WeatherScreen(
    onBack: () -> Unit
) {
    val context = LocalContext.current
    val scope = rememberCoroutineScope()
    val locationManager = remember { HuntingLocationManager(context) }
    val weatherCache = remember { WeatherCache(context) }

    var weather by remember { mutableStateOf<CachedWeather?>(null) }
    var isLoading by remember { mutableStateOf(false) }
    var errorMessage by remember { mutableStateOf<String?>(null) }
    var lastUpdated by remember { mutableStateOf<Long?>(null) }

    val moonInfo = remember { MoonPhaseCalculator.calculateMoonPhase() }

    val locationPermissionLauncher = rememberLauncherForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { granted ->
        if (granted) {
            isLoading = true
            scope.launch {
                try {
                    val location = locationManager.getCurrentLocation()
                    if (location != null) {
                        val result = weatherCache.getWeather(location.latitude, location.longitude)
                        weather = result
                        lastUpdated = result?.timestamp
                        errorMessage = null
                    } else {
                        errorMessage = "Position konnte nicht ermittelt werden"
                    }
                } catch (e: Exception) {
                    errorMessage = e.message
                }
                isLoading = false
            }
        }
    }

    // Initial load
    LaunchedEffect(Unit) {
        // Try to load cached weather first
        weather = weatherCache.getCachedWeather()
        lastUpdated = weather?.timestamp

        // Then try to refresh
        if (ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
            isLoading = true
            val location = locationManager.getCurrentLocation()
            if (location != null) {
                val result = weatherCache.getWeather(location.latitude, location.longitude)
                weather = result
                lastUpdated = result?.timestamp
            }
            isLoading = false
        }
    }

    Scaffold(
        modifier = Modifier.fillMaxSize(),
        topBar = {
            TopAppBar(
                title = {
                    Text(
                        text = "Wetter",
                        color = NightColors.onSurface
                    )
                },
                navigationIcon = {
                    IconButton(onClick = onBack) {
                        Icon(
                            Icons.Filled.ArrowBack,
                            contentDescription = "Zurueck",
                            tint = NightColors.onSurface
                        )
                    }
                },
                actions = {
                    IconButton(
                        onClick = {
                            if (ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED) {
                                isLoading = true
                                scope.launch {
                                    val location = locationManager.getCurrentLocation()
                                    if (location != null) {
                                        val result = weatherCache.getWeather(location.latitude, location.longitude, forceRefresh = true)
                                        weather = result
                                        lastUpdated = result?.timestamp
                                    }
                                    isLoading = false
                                }
                            } else {
                                locationPermissionLauncher.launch(Manifest.permission.ACCESS_FINE_LOCATION)
                            }
                        },
                        enabled = !isLoading
                    ) {
                        if (isLoading) {
                            CircularProgressIndicator(
                                modifier = Modifier.size(24.dp),
                                strokeWidth = 2.dp,
                                color = NightColors.primary
                            )
                        } else {
                            Icon(
                                Icons.Filled.Refresh,
                                contentDescription = "Aktualisieren",
                                tint = NightColors.onSurface
                            )
                        }
                    }
                },
                colors = TopAppBarDefaults.topAppBarColors(
                    containerColor = NightColors.background
                )
            )
        },
        containerColor = NightColors.background
    ) { paddingValues ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
                .padding(horizontal = 16.dp)
                .verticalScroll(rememberScrollState()),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            Spacer(modifier = Modifier.height(8.dp))

            if (weather == null && !isLoading) {
                // No weather data
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(containerColor = NightColors.surface),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(24.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Icon(
                            Icons.Filled.CloudOff,
                            contentDescription = null,
                            tint = NightColors.onBackground,
                            modifier = Modifier.size(48.dp)
                        )
                        Spacer(modifier = Modifier.height(16.dp))
                        Text(
                            text = "Keine Wetterdaten",
                            color = NightColors.onSurface,
                            fontWeight = FontWeight.Medium
                        )
                        Spacer(modifier = Modifier.height(8.dp))
                        Button(
                            onClick = {
                                locationPermissionLauncher.launch(Manifest.permission.ACCESS_FINE_LOCATION)
                            },
                            colors = ButtonDefaults.buttonColors(containerColor = NightColors.primary)
                        ) {
                            Text("Wetter abrufen")
                        }
                    }
                }
            } else if (weather != null) {
                val w = weather!!

                // Main weather card
                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(containerColor = NightColors.surface),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(20.dp),
                        horizontalAlignment = Alignment.CenterHorizontally
                    ) {
                        Text(
                            text = WeatherIconHelper.getWeatherEmoji(w.icon),
                            fontSize = 64.sp
                        )
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(
                            text = WeatherIconHelper.formatTemperature(w.temperature),
                            color = NightColors.onSurface,
                            fontSize = 48.sp,
                            fontWeight = FontWeight.Light
                        )
                        Text(
                            text = "Gefuehlt ${WeatherIconHelper.formatTemperature(w.feelsLike)}",
                            color = NightColors.onBackground,
                            fontSize = 14.sp
                        )
                        Spacer(modifier = Modifier.height(8.dp))
                        Text(
                            text = w.description.replaceFirstChar { it.uppercase() },
                            color = NightColors.onSurface,
                            fontSize = 16.sp
                        )

                        // Hunting conditions
                        val (isGood, conditionText) = WeatherIconHelper.isGoodHuntingWeather(w)
                        Spacer(modifier = Modifier.height(12.dp))
                        Text(
                            text = conditionText,
                            color = if (isGood) NightColors.success else NightColors.error,
                            fontSize = 13.sp,
                            fontWeight = FontWeight.Medium
                        )
                    }
                }

                // Wind compass rose
                SettingsSectionHeader(
                    icon = {
                        Icon(
                            Icons.Filled.Air,
                            contentDescription = null,
                            tint = NightColors.primary,
                            modifier = Modifier.size(18.dp)
                        )
                    },
                    title = "WIND"
                )

                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(containerColor = NightColors.surface),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(20.dp),
                        horizontalArrangement = Arrangement.SpaceEvenly,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        // Wind compass
                        Box(
                            modifier = Modifier.size(100.dp),
                            contentAlignment = Alignment.Center
                        ) {
                            WindCompass(windDirection = w.windDirection.toFloat())
                        }

                        Column {
                            Text(
                                text = WeatherIconHelper.formatWindSpeed(w.windSpeed),
                                color = NightColors.onSurface,
                                fontSize = 24.sp,
                                fontWeight = FontWeight.Medium
                            )
                            Text(
                                text = "aus ${w.getWindDirectionName()}",
                                color = NightColors.onBackground,
                                fontSize = 14.sp
                            )
                            Text(
                                text = WeatherIconHelper.getWindDescription(w.windSpeed),
                                color = NightColors.primary,
                                fontSize = 12.sp
                            )
                            if (w.windGust != null && w.windGust > w.windSpeed) {
                                Text(
                                    text = "Boeen: ${WeatherIconHelper.formatWindSpeed(w.windGust)}",
                                    color = NightColors.error,
                                    fontSize = 12.sp
                                )
                            }
                        }
                    }
                }

                // Details
                SettingsSectionHeader(
                    icon = {
                        Icon(
                            Icons.Filled.Info,
                            contentDescription = null,
                            tint = NightColors.primary,
                            modifier = Modifier.size(18.dp)
                        )
                    },
                    title = "DETAILS"
                )

                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(containerColor = NightColors.surface),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Column(
                        modifier = Modifier.padding(16.dp),
                        verticalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        WeatherDetailRow("Luftfeuchtigkeit", "${w.humidity}%")
                        WeatherDetailRow("Luftdruck", "${w.pressure} hPa")
                        WeatherDetailRow("Bewoelkung", "${w.cloudiness}%")
                        WeatherDetailRow("Sichtweite", "${w.visibility / 1000} km")

                        val timeFormat = SimpleDateFormat("HH:mm", Locale.GERMANY)
                        WeatherDetailRow("Sonnenaufgang", timeFormat.format(Date(w.sunrise)))
                        WeatherDetailRow("Sonnenuntergang", timeFormat.format(Date(w.sunset)))
                    }
                }

                // Moon phase
                SettingsSectionHeader(
                    icon = {
                        Icon(
                            Icons.Filled.NightsStay,
                            contentDescription = null,
                            tint = NightColors.primary,
                            modifier = Modifier.size(18.dp)
                        )
                    },
                    title = "MONDPHASE"
                )

                Card(
                    modifier = Modifier.fillMaxWidth(),
                    colors = CardDefaults.cardColors(containerColor = NightColors.surface),
                    shape = RoundedCornerShape(12.dp)
                ) {
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(20.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text(
                            text = moonInfo.phase.icon,
                            fontSize = 48.sp
                        )
                        Spacer(modifier = Modifier.width(16.dp))
                        Column {
                            Text(
                                text = moonInfo.phase.germanName,
                                color = NightColors.onSurface,
                                fontWeight = FontWeight.Medium,
                                fontSize = 16.sp
                            )
                            Text(
                                text = "Beleuchtung: ${moonInfo.illuminationPercent.toInt()}%",
                                color = NightColors.onBackground,
                                fontSize = 12.sp
                            )
                            Text(
                                text = moonInfo.activityPrediction.germanText,
                                color = NightColors.primary,
                                fontSize = 13.sp
                            )
                        }
                    }
                }

                // Last updated
                if (lastUpdated != null) {
                    val format = SimpleDateFormat("dd.MM.yyyy HH:mm", Locale.GERMANY)
                    Text(
                        text = "Zuletzt aktualisiert: ${format.format(Date(lastUpdated!!))}",
                        color = NightColors.onBackground,
                        fontSize = 11.sp,
                        modifier = Modifier.padding(top = 8.dp)
                    )

                    if (w.isExpired()) {
                        Text(
                            text = "Daten sind veraltet - bitte aktualisieren",
                            color = NightColors.error,
                            fontSize = 11.sp
                        )
                    }
                }
            }

            Spacer(modifier = Modifier.height(32.dp))
        }
    }
}

@Composable
private fun WeatherDetailRow(label: String, value: String) {
    Row(
        modifier = Modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.SpaceBetween
    ) {
        Text(
            text = label,
            color = NightColors.onBackground,
            fontSize = 14.sp
        )
        Text(
            text = value,
            color = NightColors.onSurface,
            fontSize = 14.sp,
            fontWeight = FontWeight.Medium
        )
    }
}

@Composable
private fun WindCompass(windDirection: Float) {
    val primaryColor = NightColors.primary
    val backgroundColor = NightColors.onBackground

    Canvas(modifier = Modifier.fillMaxSize()) {
        val center = Offset(size.width / 2, size.height / 2)
        val radius = size.minDimension / 2

        // Draw compass circle
        drawCircle(
            color = backgroundColor.copy(alpha = 0.3f),
            radius = radius,
            center = center,
            style = androidx.compose.ui.graphics.drawscope.Stroke(width = 2f)
        )

        // Draw cardinal directions
        val directions = listOf("N", "O", "S", "W")
        val angles = listOf(0f, 90f, 180f, 270f)

        // Draw direction markers
        for (i in 0 until 8) {
            val angle = i * 45f
            rotate(angle, center) {
                drawLine(
                    color = backgroundColor,
                    start = Offset(center.x, center.y - radius + 8),
                    end = Offset(center.x, center.y - radius + 16),
                    strokeWidth = if (i % 2 == 0) 3f else 1f
                )
            }
        }

        // Draw wind direction arrow
        rotate(windDirection, center) {
            // Arrow body
            drawLine(
                color = primaryColor,
                start = Offset(center.x, center.y + radius * 0.4f),
                end = Offset(center.x, center.y - radius * 0.6f),
                strokeWidth = 4f
            )
            // Arrow head
            drawLine(
                color = primaryColor,
                start = Offset(center.x, center.y - radius * 0.6f),
                end = Offset(center.x - 10, center.y - radius * 0.4f),
                strokeWidth = 4f
            )
            drawLine(
                color = primaryColor,
                start = Offset(center.x, center.y - radius * 0.6f),
                end = Offset(center.x + 10, center.y - radius * 0.4f),
                strokeWidth = 4f
            )
        }
    }
}

File: NoxVision-iOS/NoxVision/UI/Screens/WeatherScreen.swift
import SwiftUI

struct WeatherScreen: View {
    let onBack: () -> Void

    @EnvironmentObject var database: HuntingDatabaseManager
    @StateObject private var locationManager = HuntingLocationManager()
    @State private var weather: CachedWeather?
    @State private var isLoading = false
    @State private var error: String?
    @State private var moonInfo = MoonPhaseCalculator.calculateMoonPhase()

    var body: some View {
        VStack(spacing: 0) {
            NoxNavigationHeader(
                title: NSLocalizedString("weather", comment: ""),
                onBack: onBack,
                trailing: AnyView(
                    Button(action: fetchWeather) {
                        Image(systemName: "arrow.clockwise")
                            .foregroundColor(NoxColors.primary)
                    }
                )
            )

            if isLoading {
                Spacer()
                NoxLoadingOverlay(message: NSLocalizedString("loading", comment: ""))
                Spacer()
            } else if let weather = weather {
                ScrollView {
                    VStack(spacing: 12) {
                        // Main weather card
                        mainWeatherCard(weather)

                        // Wind compass
                        windCard(weather)

                        // Details
                        detailsCard(weather)

                        // Moon phase
                        moonPhaseCard

                        // Last updated
                        if weather.isExpired {
                            Text(NSLocalizedString("data_outdated", comment: ""))
                                .font(.system(size: 12))
                                .foregroundColor(NoxColors.warning)
                        }

                        Text(String(format: NSLocalizedString("last_updated", comment: ""),
                                    weather.timestamp.formatted(date: .abbreviated, time: .shortened)))
                            .font(.system(size: 11))
                            .foregroundColor(NoxColors.outline)
                    }
                    .padding(.horizontal, 16)
                    .padding(.bottom, 32)
                }
            } else {
                Spacer()
                VStack(spacing: 16) {
                    Image(systemName: "cloud.sun")
                        .font(.system(size: 48))
                        .foregroundColor(NoxColors.outline)
                    Text(NSLocalizedString("no_weather_data", comment: ""))
                        .foregroundColor(NoxColors.onSurfaceVariant)
                    Button(action: fetchWeather) {
                        Text(NSLocalizedString("fetch_weather", comment: ""))
                            .buttonStyle(NoxButtonStyle())
                    }
                    .buttonStyle(NoxButtonStyle())
                }
                Spacer()
            }
        }
        .onAppear {
            locationManager.requestPermission()
            weather = database.getCachedWeather()
            if weather == nil || weather?.isExpired == true {
                fetchWeather()
            }
        }
    }

    private func mainWeatherCard(_ w: CachedWeather) -> some View {
        VStack(spacing: 8) {
            Text(String(format: "%.0f°C", w.temperature))
                .font(.system(size: 48, weight: .thin))
                .foregroundColor(NoxColors.onSurface)

            Text(String(format: NSLocalizedString("feels_like", comment: ""), String(format: "%.0f°C", w.feelsLike)))
                .font(.system(size: 14))
                .foregroundColor(NoxColors.onSurfaceVariant)

            Text(w.weatherDescription.capitalized)
                .font(.system(size: 16, weight: .medium))
                .foregroundColor(NoxColors.onSurface)
        }
        .frame(maxWidth: .infinity)
        .noxCard()
    }

    private func windCard(_ w: CachedWeather) -> some View {
        VStack(spacing: 12) {
            NoxSectionHeader(title: NSLocalizedString("wind", comment: ""))

            // Wind compass
            WindCompass(direction: Double(w.windDirection))
                .frame(height: 120)

            HStack(spacing: 24) {
                VStack {
                    Text(String(format: "%.1f m/s", w.windSpeed))
                        .font(.system(size: 16, weight: .bold))
                        .foregroundColor(NoxColors.onSurface)
                    Text(String(format: NSLocalizedString("from_direction", comment: ""), w.windDirectionName))
                        .font(.system(size: 12))
                        .foregroundColor(NoxColors.onSurfaceVariant)
                }

                if let gust = w.windGust {
                    VStack {
                        Text(String(format: "%.1f m/s", gust))
                            .font(.system(size: 16, weight: .bold))
                            .foregroundColor(NoxColors.warning)
                        Text(String(format: NSLocalizedString("gusts", comment: ""), ""))
                            .font(.system(size: 12))
                            .foregroundColor(NoxColors.onSurfaceVariant)
                    }
                }
            }
        }
        .noxCard()
    }

    private func detailsCard(_ w: CachedWeather) -> some View {
        VStack(spacing: 8) {
            NoxSectionHeader(title: NSLocalizedString("details", comment: ""))
            NoxInfoRow(label: NSLocalizedString("air_humidity", comment: ""), value: "\(w.humidity)%")
            NoxInfoRow(label: NSLocalizedString("air_pressure", comment: ""), value: "\(w.pressure) hPa")
            NoxInfoRow(label: NSLocalizedString("cloud_cover", comment: ""), value: "\(w.cloudiness)%")
            NoxInfoRow(label: NSLocalizedString("visibility", comment: ""), value: String(format: "%.1f km", Double(w.visibility) / 1000))
            Divider().background(NoxColors.outlineVariant)
            NoxInfoRow(label: NSLocalizedString("sunrise", comment: ""), value: w.sunrise.formatted(date: .omitted, time: .shortened))
            NoxInfoRow(label: NSLocalizedString("sunset", comment: ""), value: w.sunset.formatted(date: .omitted, time: .shortened))
        }
        .noxCard()
    }

    private var moonPhaseCard: some View {
        VStack(spacing: 8) {
            NoxSectionHeader(title: NSLocalizedString("moon_phase", comment: ""))
            HStack(spacing: 16) {
                Text(moonInfo.phase.icon)
                    .font(.system(size: 36))
                VStack(alignment: .leading, spacing: 4) {
                    Text(moonInfo.phase.germanName)
                        .font(.system(size: 16, weight: .medium))
                        .foregroundColor(NoxColors.onSurface)
                    Text(String(format: NSLocalizedString("illumination", comment: ""), Int(moonInfo.illuminationPercent)))
                        .font(.system(size: 13))
                        .foregroundColor(NoxColors.onSurfaceVariant)
                    Text(moonInfo.activityPrediction.germanText)
                        .font(.system(size: 13))
                        .foregroundColor(activityColor)
                }
                Spacer()
            }
        }
        .noxCard()
    }

    private var activityColor: Color {
        switch moonInfo.activityPrediction {
        case .veryHigh: return NoxColors.success
        case .high: return NoxColors.primary
        case .medium: return NoxColors.warning
        case .low: return .orange
        case .veryLow: return NoxColors.error
        }
    }

    private func fetchWeather() {
        isLoading = true
        locationManager.requestPermission()

        Task {
            // Wait briefly for location
            try? await Task.sleep(nanoseconds: 1_000_000_000)

            let lat = locationManager.currentLocation?.coordinate.latitude ?? 51.1657
            let lon = locationManager.currentLocation?.coordinate.longitude ?? 10.4515

            let client = WeatherApiClient()
            let result = await client.fetchWeather(latitude: lat, longitude: lon)

            await MainActor.run {
                isLoading = false
                switch result {
                case .success(let w):
                    weather = w
                    database.saveWeather(w)
                case .failure(let err):
                    error = err.localizedDescription
                    AppLogger.shared.log("Weather fetch failed: \(err.localizedDescription)", type: .error)
                }
            }
        }
    }
}

// MARK: - Wind Compass
struct WindCompass: View {
    let direction: Double

    var body: some View {
        Canvas { context, size in
            let center = CGPoint(x: size.width / 2, y: size.height / 2)
            let radius = min(size.width, size.height) / 2 - 10

            // Draw circle
            let circlePath = Path(ellipseIn: CGRect(
                x: center.x - radius, y: center.y - radius,
                width: radius * 2, height: radius * 2
            ))
            context.stroke(circlePath, with: .color(NoxColors.outline), lineWidth: 1)

            // Draw cardinal directions
            let cardinals = ["N", "O", "S", "W"]
            for (i, label) in cardinals.enumerated() {
                let angle = Double(i) * .pi / 2 - .pi / 2
                let x = center.x + cos(angle) * (radius + 8)
                let y = center.y + sin(angle) * (radius + 8)
                context.draw(
                    Text(label).font(.system(size: 10, weight: .bold)).foregroundColor(NoxColors.onSurfaceVariant),
                    at: CGPoint(x: x, y: y)
                )
            }

            // Draw wind direction arrow
            let windAngle = direction * .pi / 180 - .pi / 2
            let arrowEnd = CGPoint(
                x: center.x + cos(windAngle) * (radius - 15),
                y: center.y + sin(windAngle) * (radius - 15)
            )

            var arrowPath = Path()
            arrowPath.move(to: center)
            arrowPath.addLine(to: arrowEnd)
            context.stroke(arrowPath, with: .color(NoxColors.primary), lineWidth: 2)

            // Arrowhead
            let headSize: CGFloat = 8
            let headAngle1 = windAngle + .pi * 0.85
            let headAngle2 = windAngle - .pi * 0.85
            var headPath = Path()
            headPath.move(to: arrowEnd)
            headPath.addLine(to: CGPoint(
                x: arrowEnd.x + cos(headAngle1) * headSize,
                y: arrowEnd.y + sin(headAngle1) * headSize
            ))
            headPath.move(to: arrowEnd)
            headPath.addLine(to: CGPoint(
                x: arrowEnd.x + cos(headAngle2) * headSize,
                y: arrowEnd.y + sin(headAngle2) * headSize
            ))
            context.stroke(headPath, with: .color(NoxColors.primary), lineWidth: 2)
        }
    }
}
